packages/eslint-plugin/src/rules/consistent-return.ts

```diff
@@ -0,0 +1,117 @@
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+export default createRule<Options, MessageIds>({
+  name: 'consistent-return',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Require `return` statements to either always or never specify values',
+      extendsBaseRule: true,
+      requiresTypeChecking: true,
+    },
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    schema: baseRule.meta.schema,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const rules = baseRule.create(context);
+    const functions: FunctionNode[] = [];
+
+    function enterFunction(node: FunctionNode): void {
+      functions.push(node);
+    }
+
+    function exitFunction(): void {
+      functions.pop();
+    }
+
+    function getCurrentFunction(): FunctionNode | null {
+      return functions[functions.length - 1] ?? null;
+    }
+
+    function isReturnPromiseVoid(
+      node: FunctionNode,
+      signature: ts.Signature,
+    ): boolean {
+      const tsNode = services.esTreeNodeToTSNodeMap.get(node);
+      const returnType = signature.getReturnType();
+      if (
+        tsutils.isThenableType(checker, tsNode, returnType) &&
+        tsutils.isTypeReference(returnType)
+      ) {
+        const typeArgs = returnType.typeArguments;
+        const hasVoid = !!typeArgs?.some(typeArg =>
+          isTypeFlagSet(typeArg, ts.TypeFlags.Void),
+        );
+        return hasVoid;
+      }
+      return false;
+    }
+
+    function isReturnVoidOrThenableVoid(node: FunctionNode): boolean {
+      const functionType = services.getTypeAtLocation(node);
+      const callSignatures = functionType.getCallSignatures();
+
+      return callSignatures.some(signature => {
+        if (node.async) {
+          return isReturnPromiseVoid(node, signature);
+        }
+        const returnType = signature.getReturnType();
+        return isTypeFlagSet(returnType, ts.TypeFlags.Void);
+      });
+    }
+
+    return {
+      ...rules,
+      FunctionDeclaration: enterFunction,
+      'FunctionDeclaration:exit'(node): void {
+        exitFunction();
+        rules['FunctionDeclaration:exit'](node);
+      },
+      FunctionExpression: enterFunction,
+      'FunctionExpression:exit'(node): void {
+        exitFunction();
+        rules['FunctionExpression:exit'](node);
+      },
+      ArrowFunctionExpression: enterFunction,
+      'ArrowFunctionExpression:exit'(node): void {
+        exitFunction();
+        rules['ArrowFunctionExpression:exit'](node);
+      },
+      ReturnStatement(node): void {
+        const functionNode = getCurrentFunction();
+        if (
+          !node.argument &&
+          functionNode &&
+          isReturnVoidOrThenableVoid(functionNode)
+        ) {
+          return;
+        }
+        rules.ReturnStatement(node);
```

Comment:
Sorry, I somehow missed that in a previous review!!

When `treatUndefinedAsUnspecified` is `true`, the base rule checks whether returned value is `undefined` or not by simply comparing node name:
https://github.com/eslint/eslint/blob/c25b4aff1fe35e5bd9d4fcdbb45b739b6d253828/lib/rules/consistent-return.js#L181-L183

I think we should check if the return value is undefined or not using type information

(both examples below have `{"treatUndefinedAsUnspecified": true}`)

So the following isn't reported:
[playground](https://deploy-preview-8289--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=GYVwdgxgLglg9mABAIwIYCcAUwA2qDmAXCnHDgKapgCUx4AJucDGOfYgN4BQiiMwibHnzVOPXonTkoIdEgZMWbcQF8uKoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQMYHsA7ZAS2QBdFDyBaaRc2aQ9KfAM3cgBpxtIAAuQCeABxS5oJUbRTwS1APQFiZStToMmLDAG1IiaNHzQemSOXoBDcgFVCAE0TsFiBwEFk95ONwkXbqyWsIgAvgC6fKEgoUA&tsconfig=&tokens=false)
```ts
function bar(flag: boolean): undefined {
  if (flag) {
    return undefined
  }
}
```

But this one is reported:
[playground](https://deploy-preview-8289--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gNwFzwFcVQAzASxRGACgSiwMzV4AjWAChOgHN8WkkEEFBQBKfEVIUq8AN7V48MiXiceouQsXw4GAjBQ4tAX2rGgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQMYHsA7ZAS2QBdFDyBaaRc2aQ9KfAM3cgBpxtIAAuQCeABxS5oJUbRTwS1APQFiZStToMmLDAG1IiaNHzQemSOXoBDcgFVCAE0TsFiBwEFk95ONwkXbqyWsIgAvgC6fKEgoUA&tsconfig=&tokens=false)
```ts
declare const v: undefined
function bar(flag: boolean): undefined {
  if (flag) {
    return v
  }
}
```

---

Not sure what's the best way to fix this, because this logic is hidden in the base rule. The only thing I can think of is to somehow fake the argument of return statement node before passing it to the `ReturnStatement` base rule listener, like here:

https://github.com/typescript-eslint/typescript-eslint/blob/bbc6770c05d3521a155b8a73ebca963c4d90bf25/packages/eslint-plugin/src/rules/no-extra-parens.ts#L137-L143